File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size(); 
        int ind = -1, diff = INT_MAX; 
        for(int i=n-2; i>=0; i--) {
            if(nums[i] < nums[i+1]) {
                ind = i; 
                break;
            }
        }

        if(ind == -1) {
            reverse(nums.begin(), nums.end()); 
            return;
        }
        int jnd = 0;
        for(int i=n-1; i>=0; i--) {
            if(nums[i] > nums[ind]) {
                jnd = i;
                break;
            }
        }
        swap(nums[ind], nums[jnd]);
        reverse(nums.begin()+ind+1, nums.end());
    }
};

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    public void reverse(int[] nums, int i, int j) {
        while(i < j) swap(nums, i++, j--);
    }
    
    public void nextPermutation(int[] nums) {
        int ind = -1; 
        for(int i=nums.length-2; i>=0; i--) {
            if(nums[i] < nums[i+1]) {
                ind = i;
                break;
            }
        }
        
        if(ind == -1) {
            reverse(nums, 0, nums.length-1);
            return;
        }
        
        int jnd = -1;
        for(int i=nums.length-1; i>=0; i--) {
            if(nums[i] > nums[ind]) {
                jnd = i;
                break;
            }
        }
        swap(nums, ind, jnd);
        reverse(nums, ind+1, nums.length-1);
    }
}